azure-pipelines.yml

```diff
@@ -1,203 +1,73 @@
-
-  - script: |
-      python -c '
-      from pathlib import Path
-      import shutil
-
-      from_folder = Path("$(Build.StagingDirectory)")
-      destination_folder = Path("$(System.DefaultWorkingDirectory)") / ".tox"
-      destination_folder.mkdir()
-      for coverage_file in from_folder.glob("*/.coverage"):
-          destination = destination_folder / f".coverage.{coverage_file.parent.name[9:]}"
-          print(f"{coverage_file} copy to {destination}")
-          shutil.copy(str(coverage_file), str(destination))'
-    displayName: move coverage files into .tox
-
-  - script: "python -m pip install -U pip setuptools --user"
-    displayName: upgrade pip
-
-  - script: 'python -m pip install . -U --user'
-    displayName: install tox - eat our own dog food
-
-  - script: 'python -m tox -e py --sdistonly'
-    displayName: generate version.py
-
-  - script: 'python -m tox -e coverage'
-    displayName: create coverag report via tox
-
-  - task: PublishCodeCoverageResults@1
-    displayName: publish overall coverage report to Azure
-    inputs:
-      codeCoverageTool: 'Cobertura'
-      summaryFileLocation: "$(System.DefaultWorkingDirectory)/.tox/coverage.xml"
-      failIfCoverageEmpty: true
-
-  - script: |
-      curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./cc-test-reporter && \
-      chmod +x ./cc-test-reporter && \
-      python -c '
-      from xml.etree import ElementTree as et
-      from pathlib import Path
-      import subprocess
-
-      from_folder = Path("$(Build.StagingDirectory)")
-      for counter, coverage_file in enumerate(from_folder.glob("*/coverage.xml")):
-          key = coverage_file.parent.name[9:]
-          print(f"{counter}) {coverage_file}")
-          try:
-              cmd = ["$(System.DefaultWorkingDirectory)/cc-test-reporter", "format-coverage",
-                     str(coverage_file),
-                     "-d", "-t", "coverage.py",
-                     "-o", f"$(Build.StagingDirectory)/code-climate.{key}.json"]
-              print(f"\t{cmd}")
-              log = subprocess.check_output(cmd, stderr=subprocess.STDOUT, universal_newlines=True)
-              code = 0
-          except subprocess.CalledProcessError as exception:
-              log, code = exception.output, exception.returncode
-          finally:
-              print(code, log, "\n", sep="\n")' && \
-      ./cc-test-reporter sum-coverage -d --output - \
-          --parts $(ls -1 $(Build.StagingDirectory)/code-climate.*.json | wc -l) \
-          $(Build.StagingDirectory)/code-climate.*.json | \
-      ./cc-test-reporter -d -r d24f105984ab5e087773a21b8668acb0b36cb8311fc2637f78a2d9451e531e08 \
-          upload-coverage --input -
-    displayName: publish code climate
-    condition: succeededOrFailed()
+- template: run-tox-env.yml@tox-azure
+  parameters:
+    jobs:
+      check:
+        py: '3.7'
+        toxenvs:
+        ? fix_lint
+        ? docs
+        ? package_description
```

Comment:
hmmm! how does this syntax work?

Reply:
yaml sets?

Reply:
wow, I've *literal*ly never seen this

Reply:
apparently it's a "key indicator" so it makes mappings with no values?

```pycon
>>> yaml.safe_load('''\
... x:
...   ? a
...   ? b
...   ? c
... ''')
{'x': {'a': None, 'b': None, 'c': None}}
```

Reply:
yes per standard defintion

Reply:
changed back to array now 👍